cmd/kaspaminer/mineloop.go

```diff
@@ -0,0 +1,202 @@
+		default:
+			msgBlock.Header.Nonce = i
+			hash := msgBlock.BlockHash()
+			if daghash.HashToBig(hash).Cmp(targetDifficulty) <= 0 {
+				foundBlock <- block
+				return
+			}
+		}
+	}
+
+}
+
+func getBlockTemplate(client *minerClient, longPollID string) (*rpcmodel.GetBlockTemplateResult, error) {
+	return client.GetBlockTemplate([]string{"coinbasetxn"}, longPollID)
+}
+
+func templatesLoop(client *minerClient, newTemplateChan chan *rpcmodel.GetBlockTemplateResult, errChan chan error, stopChan chan struct{}) {
+	longPollID := ""
+	getBlockTemplateLongPoll := func() {
+		if longPollID != "" {
+			log.Infof("Requesting template with longPollID '%s' from %s", longPollID, client.Host())
+		} else {
+			log.Infof("Requesting template without longPollID from %s", client.Host())
+		}
+		template, err := getBlockTemplate(client, longPollID)
+		if nativeerrors.Is(err, rpcclient.ErrResponseTimedOut) {
+			log.Infof("Got timeout while requesting template '%s' from %s", longPollID, client.Host())
+			return
+		} else if err != nil {
+			errChan <- errors.Errorf("Error getting block template from %s: %s", client.Host(), err)
+			return
+		}
+		if template.LongPollID != longPollID {
+			log.Infof("Got new long poll template: %s", template.LongPollID)
+			longPollID = template.LongPollID
+			newTemplateChan <- template
+		}
+	}
+	getBlockTemplateLongPoll()
+	for {
+		select {
+		case <-stopChan:
+			close(newTemplateChan)
+			return
+		case <-client.onBlockAdded:
+			getBlockTemplateLongPoll()
+		case <-time.Tick(500 * time.Millisecond):
+			getBlockTemplateLongPoll()
+		}
+	}
+}
+
+func solveLoop(newTemplateChan chan *rpcmodel.GetBlockTemplateResult, foundBlock chan *util.Block, errChan chan error) {
+	var stopOldTemplateSolving chan struct{}
+	for template := range newTemplateChan {
+		if stopOldTemplateSolving != nil {
+			close(stopOldTemplateSolving)
+		}
+		stopOldTemplateSolving = make(chan struct{})
+		block, err := parseBlock(template)
+		if err != nil {
+			errChan <- errors.Errorf("Error parsing block: %s", err)
+			return
+		}
+
+		go solveBlock(block, stopOldTemplateSolving, foundBlock)
+	}
+	if stopOldTemplateSolving != nil {
+		close(stopOldTemplateSolving)
+	}
+}
+
+func mineNextBlock(client *minerClient, foundBlock chan *util.Block, templateStopChan chan struct{}, errChan chan error) {
+	newTemplateChan := make(chan *rpcmodel.GetBlockTemplateResult)
+	go templatesLoop(client, newTemplateChan, errChan, templateStopChan)
+	go solveLoop(newTemplateChan, foundBlock, errChan)
+}
+
+func handleFoundBlock(client *minerClient, block *util.Block) error {
+	log.Infof("Found block %s with parents %s. Submitting to %s", block.Hash(), block.MsgBlock().Header.ParentHashes, client.Host())
+
+	err := client.SubmitBlock(block, &rpcmodel.SubmitBlockOptions{})
+	if err != nil {
+		return errors.Errorf("Error submitting block %s to %s: %s", block.Hash(), client.Host(), err)
+	}
+	return nil
+}
+
+func mineLoop(client *minerClient, numberOfBlocks uint64) error {
```

Comment:
Please arrange the functions in some sort of logical/chronological order. E.g. mineLoop first, mineNextBlock seconds, templatesLoop third, etc.